File: jars/src/data/weapons/ApexDartgunMuzzleFlash.java
package data.weapons;

import com.fs.starfarer.api.combat.*;
import org.lazywizard.lazylib.MathUtils;
import org.lwjgl.util.vector.Vector2f;

import java.awt.*;

public class ApexDartgunMuzzleFlash implements OnFireEffectPlugin, EveryFrameWeaponEffectPlugin
{
    // smoke vents
    public static final float PARTICLE_SIZE_MIN = 10f;
    public static final float PARTICLE_SIZE_MAX = 30f;
    public static final float PARTICLE_DURATION = 0.33f;
    public static final float PARTICLE_SPREAD = 30f;
    public static final float PARTICLE_VEL_MIN = 25f;
    public static final float PARTICLE_VEL_MAX = 100f;
    public static final float PARTICLE_MIN_DISTANCE = 0f;
    public static final float PARTICLE_MAX_DISTANCE = 20f;
    public static final int PARTICLE_NUM = 20;
    public static final Color PARTICLE_COLOR = new Color(50,50,50,100);
    public static final float[] angles = {0};

    // muzzle flash
    @Override
    public void onFire(DamagingProjectileAPI proj, WeaponAPI weapon, CombatEngineAPI engine)
    {

        for (float baseAngle : angles)
        {
            for (int i = 0; i < PARTICLE_NUM; i++)
            {
                float actualFacing = weapon.getCurrAngle() + baseAngle;
                float arcPoint = MathUtils.getRandomNumberInRange(actualFacing-(PARTICLE_SPREAD/2f), actualFacing+(PARTICLE_SPREAD/2f));
                Vector2f velocity = MathUtils.getPointOnCircumference(weapon.getShip().getVelocity(), MathUtils.getRandomNumberInRange(PARTICLE_VEL_MIN, PARTICLE_VEL_MAX), arcPoint);

                //Gets a spawn location in the cone, depending on our offsetMin/Max
                Vector2f spawnLocation = MathUtils.getPointOnCircumference(proj.getLocation(), MathUtils.getRandomNumberInRange(PARTICLE_MIN_DISTANCE, PARTICLE_MAX_DISTANCE), arcPoint);
                engine.addSmoothParticle(spawnLocation, velocity, MathUtils.getRandomNumberInRange(PARTICLE_SIZE_MIN, PARTICLE_SIZE_MAX), 1f, PARTICLE_DURATION, PARTICLE_COLOR);
            }
        }
    }

    @Override
    public void advance(float v, CombatEngineAPI combatEngineAPI, WeaponAPI weaponAPI)
    {
        // do nothing
    }
}

File: jars/src/data/weapons/ApexDartgunMuzzleFlash.kt
package data.weapons

import com.fs.starfarer.api.combat.CombatEngineAPI
import com.fs.starfarer.api.combat.DamagingProjectileAPI
import com.fs.starfarer.api.combat.OnFireEffectPlugin
import com.fs.starfarer.api.combat.WeaponAPI
import utils.ApexGenericMuzzleFlash
import java.awt.Color

class ApexDartgunMuzzleFlash: ApexGenericMuzzleFlash(), OnFireEffectPlugin
{
    // commented stuff is
    //var lastProj: DamagingProjectileAPI? = null
    //var adjust = false
    override fun onFire(proj: DamagingProjectileAPI?, weapon: WeaponAPI?, engine: CombatEngineAPI?)
    {
        proj ?: return; weapon ?: return; engine ?: return
        super.spawnSmoke(proj, weapon, engine)

        // TODO: smoke vfx
    }

    override val USED_IDS = listOf(
        "smoke1",
        "smoke2",
        "smoke3",
        "flash",
        "flashcore"
    )
    override val PARTICLE_TYPE = mapOf(
        "flash" to "HIT",
        "flashcore" to "HIT"
    )
    override val PARTICLE_COUNT = mapOf(
        "smoke3" to 10,
        "flash" to 1,
        "flashcore" to 1
    )
    override val PARTICLE_COLOR = mapOf(
        "smoke1" to Color(140,130,120,128),
        "smoke2" to Color(90, 80, 80, 118),
        "flash" to Color(255, 170, 60, 255),
        "flashcore" to Color(255, 240, 230, 255)
    )
    override val PARTICLE_OFFSET_MAX = mapOf(
        "smoke1" to 30f,
        "smoke2" to 20f,
        "smoke3" to 5f,
        "flash" to 0f,
        "flashcore" to 0f
    )
    override val PARTICLE_VELOCITY_MAX = mapOf(
        "smoke3" to 20f,
        "flash" to 0f,
        "flashcore" to 0f
    )
    override val PARTICLE_DURATION_MIN = mapOf(
        "flash" to 0.1f,
        "flashcore" to 0.05f
    )
    override val PARTICLE_DURATION_MAX = mapOf(
        "flash" to 0.1f,
        "flashcore" to 0.05f
    )
    override val PARTICLE_SIZE_MIN = mapOf(
        "flash" to 70f,
        "flashcore" to 50f
    )
    override val PARTICLE_SIZE_MAX = mapOf(
        "flash" to 70f,
        "flashcore" to 50f
    )
    override val PARTICLE_ARC = mapOf(
        "smoke3" to 160f
    )
}